### PR TITLE
Append indexes to duplicate expression names

### LINKED ISSUE #12104: If a custom column name is reused it replaces the old column silently
**Describe the bug**
It looks like we're silently replacing custom columns if you use the same name. Admittedly I'm not sure how much this is likely to happen in the real world but it's still not great.

**To Reproduce**
1. Create a custom column using any expression with a name, say `Test`
2. Create another custom, again using any expression you want, also name this one `Test`
3. The second custom column replaces the first one without any sort of feedback.

**Expected behavior**
We should either not allow you to reuse names and validate the name is unique before you add a column or append a 1,2, etc if we're ok with name reuse.

### PR DESCRIPTION
**Pull request by kulyk**
Fixes #12104

In the notebook editor, if you have an expression called e.g. `EXP` and you add another expression called `EXP`, it will silently replace the first expression, which can be confusing. This PR changes that behavior, so we append indexes to duplicate names. So if we're adding `EXP` and an `EXP` expression already exists, the new one will be added as `EXP (1)`, `EXP (2)`, etc.

### To Verfy

1. Ask a question > Custom question > Orders
2. Click "Custom column"
3. Add any expression, let's say it's called "EXP", but any name should work
4. Click the plus icon near the "EXP" notebook cell
5. Add any expression and use the same name as from step 3, repeat it a few times
6. You should see columns like "EXP", "EXP (1)", "EXP (2)", etc.
7. Other names for expressions should work fine

### Demo

**Before**

https://user-images.githubusercontent.com/17258145/135498404-943165d9-8890-4419-a05a-be04d6c4ac1a.mov

**After**

https://user-images.githubusercontent.com/17258145/135498422-d831ce43-e4f1-4e66-a585-3dc69b91d4ff.mov

### CODE REVIEW

**Review by daltojohnso — APPROVED** [review]
looks good!

**Review by flamber — COMMENTED** [review]
Great and simple idea, but seeing some tiny strangeness, which are likely rare, but just wanted to point them out:

https://user-images.githubusercontent.com/1447303/135603021-e96fe1d1-d704-40f8-b3f9-cd2f83cacea3.mp4

**Review by kdoh — COMMENTED** [review]
👍  Works exactly like I was imagining. Nice work. It'd be great if we could figure out some fixes for the issues Flamber found with renaming and removing other expressions as well though.

**Comment by kulyk** [comment]
Thanks for your review @flamber. I addressed the issues you mentioned, would appreciate one more look on the branch

File: frontend/test/metabase/scenarios/question/notebook.cy.spec.js [hunk 1]
```diff
@@ -133,6 +133,55 @@ describe("scenarios > question > notebook", () => {
     });
   });
 
+  it("should append indexes to duplicate custom expression names", () => {
```
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/question/notebook.cy.spec.js:136]:
```suggestion
  it("should append indexes to duplicate custom expression names (metabase#12104)", () => {
```

File: frontend/test/metabase/scenarios/question/notebook.cy.spec.js [hunk 2]
```diff
@@ -133,6 +133,55 @@ describe("scenarios > question > notebook", () => {
...
+    cy.findByTestId("step-expression-0-0").within(() => {
+      cy.icon("add").click();
+    });
+    popover()
+      .findByText("Category")
+      .click();
+    cy.findByPlaceholderText("Something nice and descriptive")
+      .click()
+      .type("EXPR");
+    cy.button("Done").click();
```
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/question/notebook.cy.spec.js:158]: Nit: This sequence is repeating enough times. What do you think about extracting it into a helper function to make this test more succint?

**Pull request merged**